### ISSUE TITLE
Trampolined StreamT should not stack-overflow on foldLeft

### ISSUE THREAD

**Issue by TomasMikula** [scalaz/scalaz#1123]
`StreamT#foldLeft` overflows the stack, even when used on a `Trampoline`d stream. This is due to taking arguments by-name. I propose to fix this by changing the signature from

```scala
def foldLeft[B](z: => B)(f: (=> B, => A) => B)(implicit M: Monad[M]): M[B]
```

to

```scala
def foldLeft[B](z: B)(f: (B, A) => B)(implicit M: Monad[M]): M[B]
```

**Comment by TomasMikula** [comment]
A failing test case is in #1124.

**Comment by TomasMikula** [comment]
Implementation of the proposed fix is in #1125.

**Comment by TomasMikula** [comment]
Fixed by #1125.
